R/partykit_skins.R

```diff
@@ -0,0 +1,303 @@
+      breaks <- cbind(c(-Inf, split$breaks), c(split$breaks, Inf))
+      sbreak <- breaks[index == whichkid, ]
+      # right <- partykit::right_split(split)
+      right <- split$right
+      srule <- c()
+      if (is.finite(sbreak[1])) {
+        srule <- c(srule, paste(svar, ifelse(right, ">", 
+                                             ">="), sbreak[1]))
+      }
+      if (is.finite(sbreak[2])) { 
+        srule <- c(srule, paste(svar, ifelse(right, "<=", 
+                                             "<"), sbreak[2]))
+      }
+      srule <- paste(srule, collapse = " & ")
+    }
+    rule <<- c(rule, srule)
+    return(recFun(node[[whichkid]]))
+  }
+  # node <- recFun(partykit::node_party(x))
+  node <- recFun(x$node)
+  paste(rule, collapse = " & ")
+}
+
+#####
+# Wrappers for ctree which returns the needed output to get rules
+
+#' @importFrom stats model.weights na.pass
+ctree_setup <- function(
+  formula, data, weights, subset, na.action = na.pass, 
+  control = ctree_control(...), ytrafo = NULL, scores = NULL, 
+  ...
+){
+  if (missing(data)) 
+    data <- environment(formula)
+  mf <- match.call(expand.dots = FALSE)
+  m <- match(c("formula", "data", "subset", "weights", "na.action"), 
+             names(mf), 0)
+  mf <- mf[c(1, m)]
+  formula <- Formula::Formula(formula)
+  mf$formula <- formula
+  mf$drop.unused.levels <- FALSE
+  mf$na.action <- na.action
+  mf[[1]] <- quote(stats::model.frame)
+  mf <- eval(mf, parent.frame())
+  response <- names(Formula::model.part(formula, mf, lhs = 1))
+  weights <- model.weights(mf)
+  dat <- mf[, colnames(mf) != "(weights)"]
+  if (!is.null(scores)) {
+    for (n in names(scores)) {
+      sc <- scores[[n]]
+      if (is.ordered(dat[[n]]) && nlevels(dat[[n]]) == 
+          length(sc)) {
+        attr(dat[[n]], "scores") <- as.numeric(sc)
+      }
+      else {
+        warning("scores for variable ", sQuote(n), " ignored")
+      }
+    }
+  }
+  if (is.null(weights)) 
+    weights <- rep(1, nrow(mf))
+  storage.mode(weights) <- "integer"
+  nvar <- sum(!(colnames(dat) %in% response))
+  
+  control$cfun <- eval(bquote(function(...) {
+    if (.(control$teststat == "quad")) 
+      p <- .pX2(..., pval = .(control$testtype != "Teststatistic"))
+    if (.(control$teststat == "max"))
+      p <- .pmaxT(..., pval = .(control$testtype != "Teststatistic"))
+    names(p) <- c("statistic", "p.value")
+    if (.(control$testtype == "Bonferroni")) 
+      p["p.value"] <- p["p.value"] * .(min(nvar, control$mtry))
+    crit <- p["statistic"]
+    if (.(control$testtype != "Teststatistic")) 
+      crit <- p["p.value"]
+    c(crit, p)
+  }))
+  environment(control$cfun) <- environment(ctree)
+  
+  list(dat = dat, response = response, weights = weights, 
+       control = control, ytrafo = ytrafo)
+}
+
+ctree_minmal <- function (
+  dat, response, control, ytrafo, ...) 
+{
+  .ctree_fit <- with(environment(ctree), .ctree_fit)
```

Comment:
New partykit version has no function .ctree_fit anymore 

They have developed a faster version now (see https://r-forge.r-project.org/scm/viewvc.php/pkg/partykit/R/ctree.R?view=markup&revision=1098&root=partykit); haven't figured out the name of the new function.